File: Python/1965_상자넣기.py
if __name__ == '__main__':
	n = int(input())
	box = list(map(int,input().split()))
	
	max_cnt = 0  # 최대 상자의 개수를 담는 변수
	agg_box = [1]*n # 박스가 몇 개의 박스를 포함하고 있는지(본인도 포함되기 때문에 1로 초기화)
	for i in range(1,n):
		tmp = 0 # 앞에 있는 박스들의 값 중 최대값을 담을 변수
		for j in range(i):
			if box[i] > box[j] and tmp < agg_box[j]: # 현재 박스보다 작고 지금까지 비교한 박스들 중 포함된 박스의 값이 가장 큰 값이라면
				tmp = agg_box[j]

		agg_box[i] = tmp + 1 # 앞에 있는 박스들의 값중 최대값 + 1(본인 포함)
		if agg_box[i] > max_cnt: # 그 중 최대 값을 구하기 위함
			max_cnt = agg_box[i]
			
	print(max_cnt)

File: C++/1965_상자넣기.cpp
#include <iostream>
using namespace std;

int main() {
	int N;
	cin >> N;
	
	int box[N]; // 박스 크기 담은 배열
	for(int n = 0; n < N; n++){
		cin >> box[n];
	}
	
	int arr[N], max_box = 0; // arr은 현재 위치에서 담긴 박스의 최대 개수, max_box는 담을 수 있는 박스의 최대값
	arr[0] = 1;
	for(int i = 1; i < N; i++){
		int tmp = 0; // 현재 박스의 앞 쪽에 있는 박스들 중 담겨 있는 박스 수의 최대값 구하기 위함
		for(int j = 0; j < i; j++){ // 현재 박스의 앞 쪽에 있는 박스들 확인
			if(box[j] < box[i] && arr[j] > tmp){ // 앞 쪽 박스가 현재 박스보다 작고 비교했던 박스들보다 더 많은 박스 품고 있을 때
				tmp = arr[j];
			}
		}
		arr[i] = tmp + 1; // 현재 박스까지 포함해야하기 때문에 +1
		if(arr[i] > max_box){
			max_box = arr[i]; // 그 중 가장 큰 값을 max_box에 저장
		}
	}
	
	cout << max_box;
	
	return 0;
}
